### PR TITLE
Feature/storage/data lake rename sas bug

### LINKED ISSUE #11284: [Storage][DataLake] PathClient.Rename() appears to not function correctly with SAS
Reported by @blueww

### PR DESCRIPTION
**Pull request by seanmcc-msft**
Resolves https://github.com/Azure/azure-sdk-for-net/issues/11284

### CODE REVIEW

File: sdk/storage/Azure.Storage.Files.DataLake/tests/DirectoryClientTests.cs [hunk 1]
```diff
@@ -479,6 +480,51 @@ public async Task RenameAsync()
...
+            {
+                FileSystemName = test.FileSystem.Name,
+                Path = null,
+                Protocol = SasProtocol.None,
+                StartsOn = Recording.UtcNow.AddHours(-2),
+                ExpiresOn = Recording.UtcNow.AddHours(+2),
+                IPRange = new SasIPRange(IPAddress.None, IPAddress.None)
+            };
+            sasBuilder.SetPermissions(DataLakeSasPermissions.All);
+            SasQueryParameters destSasQueryParameters = sasBuilder.ToSasQueryParameters(GetStorageSharedKeyCredentials());
```
Comment (blueww) [inline sdk/storage/Azure.Storage.Files.DataLake/tests/DirectoryClientTests.cs:516]: If we have created Dest Dir Client, can we user "destDataLakeDirectoryClient.Uri.PathAndQuery"?
Do you see any potential issue? 
If not, I will use this in PowerShell.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.DataLake/tests/DirectoryClientTests.cs:516]: Could you rephrase the question?

Comment (blueww) [reply sdk/storage/Azure.Storage.Files.DataLake/tests/DirectoryClientTests.cs:516]: I mean if I create a dest Directory client object, then get "destDataLakeDirectoryClient.Uri.PathAndQuery" and input to rename API as dest. 

In my testing, I see "PathAndQuery" will contain both path and the SAS, does it works? Not only for SAS, but also for other credential like sharedkey, Oauth.

**Comment by seanmcc-msft** [comment]
Consider adding another Rename overload taking file system + path for the destination path.

File: sdk/storage/Azure.Storage.Files.DataLake/tests/DirectoryClientTests.cs (hunk 1 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.DataLake/tests/DirectoryClientTests.cs:516]: Currently, destination file system and destination path are 2 separate parameters.

A client created by Rename() using SAS can only be used with SAS going forward, since it was created with SAS, and clients are immutable.

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePathClient.cs [hunk 2]
```diff
@@ -321,6 +326,7 @@ internal DataLakePathClient(Uri pathUri, HttpPipelinePolicy authentication, Data
             _version = options.Version;
             _clientDiagnostics = new ClientDiagnostics(options);
             _blockBlobClient = BlockBlobClientInternals.Create(_blobUri, _pipeline, Version.AsBlobsVersion(), _clientDiagnostics);
+            _options = options;
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePathClient.cs:329]: This is kind of problematic.  The pipeline is designed to be immutable but the client options are not.  We're supposed to cache any values needed from the options when creating the pipeline rather than hanging on to the options object directly.  We already do that for things like the diagnostics, version, etc.  Is there any reason we can't keep doing that?

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePathClient.cs [hunk 3]
```diff
@@ -1550,12 +1556,33 @@ internal virtual void SetNameFieldsIfNull()
...
+                    destUriBuilder.DirectoryOrFilePath = destinationPath;
+
+                    // We will get sas with destPath, if it was provided by the user.
+                    destUriBuilder.Sas = null;
+
+                    // Build destPathClient
+                    DataLakePathClient destPathClient;
+                    if (destUriBuilder.Sas != null)
+                    {
+                        destPathClient = new DataLakePathClient(destUriBuilder.ToUri(), _options);
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePathClient.cs:1580]: I don't quite follow why we're not passing the pipeline here.  It should be the same pipeline if you were using Sas before the rename.

**Review by tg-msft — COMMENTED** [review]
What happens if I pass a SAS URI as my client endpoint and then also use a SharedKeyCredential?  Did that work before this change (because I don't think it would after if you're not threading the pipeline through to the renamed client)?

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePathClient.cs (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePathClient.cs:1580]: Fixed.  Dumb mistake on my part.

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePathClient.cs (hunk 2 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.DataLake/src/DataLakePathClient.cs:329]: Fixed, we didn't need this.

**Comment by seanmcc-msft** [comment]
> What happens if I pass a SAS URI as my client endpoint and then also use a SharedKeyCredential? Did that work before this change (because I don't think it would after if you're not threading the pipeline through to the renamed client)?

You can't use both SAS and SharedKeyCredential at the same time.  I fixed the issue with not passing the pipeline in, that was a dumb mistake on my part.

**Review by tg-msft — APPROVED** [review]
Looks good
